lib/charms/kratos/v0/kubernetes_network_policies.py

```diff
@@ -0,0 +1,160 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class NetworkPoliciesHandlerError(Exception):
+    """Applying the network policies failed."""
+
+
+Port = Union[str, int]
+IngressPolicyDefinition = Tuple[Port, List[Relation]]
+
+
+class KubernetesNetworkPoliciesHandler:
+    """A helper class for managing kubernetes network policies."""
+
+    def __init__(self, charm: CharmBase) -> None:
+        self._charm = charm
+        self.client = Client(field_manager=charm.app.name, namespace=charm.model.name)
+
+    @property
+    def policy_name(self) -> str:
+        """The default policy name that will be created."""
+        return f"{self._charm.app.name}-network-policy"
+
+    def apply_ingress_policy(
+        self, policies: IngressPolicyDefinition, name: Optional[str] = None
+    ) -> None:
+        """Apply an ingress network policy about a related application.
+
+        Policies can be defined for multiple ports at once to allow ingress traffic
+        from related applications
+
+        If no policies are defined then all ingress traffic will be denied.
+
+        Example usage:
+
+            policies = [("admin", [admin_ingress_relation]), (8080, [public_ingress_relation])]
+            network_policy_handler.apply_ingress_policy(policies)
+        """
+        if not name:
+            name = self.policy_name
+
+        ingress = []
+        for port, relations in policies:
+            selectors = [
+                NetworkPolicyPeer(
+                    podSelector=LabelSelector(
+                        matchLabels={"app.kubernetes.io/name": relation.app.name}
+                    )
+                )
+                for relation in relations
+                if relation and relation.app
+            ]
+            ingress.append(
+                NetworkPolicyIngressRule(
+                    from_=selectors,
+                    ports=[NetworkPolicyPort(port=port, protocol="TCP")],
+                ),
+            )
+
+        policy = NetworkPolicy(
+            metadata=ObjectMeta(name=name),
+            spec=NetworkPolicySpec(
+                podSelector=LabelSelector(
+                    matchLabels={
+                        "app.kubernetes.io/name": self._charm.app.name,
+                    }
+                ),
+                policyTypes=["Ingress", "Egress"],
+                ingress=ingress,
+            ),
+        )
+
+        try:
+            self.client.apply(
+                policy,
+                namespace=self._charm.model.name,
+            )
+        except ApiError as e:
+            if e.status.code == 403:
+                msg = f"Kubernetes resources patch failed: `juju trust` this application. {e}"
+            else:
+                msg = f"Kubernetes resources patch failed: {e}"
+            logger.error(msg)
+            raise NetworkPoliciesHandlerError()
+
+    def delete_network_policy(self, name: Optional[str] = None) -> None:
```

Comment:
We should standardise on naming, we use `ingress policies` and `network policies` interchangeably.